pika/adapters/asyncio_connection.py

```diff
@@ -0,0 +1,476 @@
+        # to the channel.
+        if name in self.WRAPPED_METHODS:
+            return self.__wrap_channel_method(name)
+        return getattr(self.__channel, name)
+
+
+class IOLoopAdapter:
+    """An adapter providing Pika's IOLoop interface using a asyncio IOLoop.
+
+    Accepts a AsyncioConnection object and a asyncio IOLoop.
+
+    """
+
+    def __init__(self, connection, loop):
+        self.connection = connection
+        self.loop = loop
+        self.started = False
+
+    def add_timeout(self, deadline, callback_method):
+        """Add the callback_method to the IOLoop timer to fire after deadline
+        seconds. Returns a handle to the timeout. Do not confuse with
+        Tornado's timeout where you pass in the time you want to have your
+        callback called. Only pass in the seconds until it's to be called.
+
+        :param int deadline: The number of seconds to wait to call callback
+        :param method callback_method: The callback method
+        :rtype: asyncio.Handle
+
+        """
+        return self.loop.call_later(deadline, callback_method)
+
+    def remove_timeout(self, call):
+        """Remove a call
+
+        :param asyncio.Handle call: The call to cancel
+
+        """
+        call.cancel()
+
+    def stop(self):
+        # Guard against stopping the loop multiple times
+        if not self.started:
+            return
+        self.started = False
+        self.loop.stop()
+
+    def start(self):
+        # Guard against starting the loop multiple times
+        if self.started:
+            return
+        self.started = True
+        self.loop.run_forever()
+
+    def remove_handler(self, _):
+        # The fileno is irrelevant, as it's the connection's job to provide it
+        # to the loop when asked to do so. Removing the handler from the
+        # ioloop is removing it from the loop in acyncio's parlance.
+        self.loop.remove_reader(self.connection.fileno())
+        self.loop.remove_writer(self.connection.fileno())
+
+    def update_handler(self, _, event_state):
+        # Same as in remove_handler, the fileno is irrelevant. First remove the
+        # connection entirely from the reactor, then add it back depending on
+        # the event state.
+        self.loop.remove_reader(self.connection.fileno())
+        self.loop.remove_writer(self.connection.fileno())
+
+        if event_state & self.connection.READ:
+            self.loop.add_reader(self.connection.fileno(), self._handle_read)
+
+        if event_state & self.connection.WRITE:
+            self.loop.add_writer(self.connection.fileno(), self._handle_write)
+
+
+class AsyncioProtocolConnection(base_connection.BaseConnection):
+    """A hybrid between a Pika Connection and a acyncio Protocol. Allows using
+    asyncio's non-blocking loop.create_connection method for connecting to the
+    server.
+
+    It has one caveat: AcyncioProtocolConnection objects have a ready
```

Comment:
This adapter use different approach then  SelectConnection or TornadoConnection, so i suppose existing acceptance testing logic may not fit to him